### PR TITLE
[SPARK-44850][CONNECT] Heartbeat in scala's Spark Connect

### PR DESCRIPTION
**Pull request by cdkrot**
### What changes were proposed in this pull request?
Implement a heartbeat for spark connect. This works by maintaining the number of open queries and sending "keep-alive" requests from a separate thread if at least one such query is running

### Why are the changes needed?

We notice that clients who wish to execute long requests (specifically this happens to ExecutePlanRequests taking 1 hour or more), often face disconnects by intermediate proxy layers, such as those in common cloud providers. Apparently more standard ways to resolve this such as grpc's heartbeat don't help.

### Does this PR introduce _any_ user-facing change?
This change proposes to enable heartbeat by default

### How was this patch tested?

UT and E2E coverage. It was also separately verified that 1 hour 10 min query worked fine, i.e.

```
test("Alice") {
    val n: Long = 80e13.toLong

    val res = spark.range(n).count()

    assert(res == n)
}
```

### CODE REVIEW

File: connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/SparkSessionE2ESuite.scala [hunk 1]
```diff
@@ -106,6 +108,110 @@ class SparkSessionE2ESuite extends RemoteSparkSession {
...
+    }
+
+    eventually(timeout(10.seconds), interval(100.millis)) {
+      spark.interruptAll()
+      assert(spark.client.heartbeatLevel == 0)
+    }
+  }
+
+  test("Test heartbeat gets executed") {
+    // Importing here to avoid extra edge changes in imports section
```
Comment (hvanhovell) [inline connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/SparkSessionE2ESuite.scala]: Remove this?

Comment (cdkrot) [reply connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/SparkSessionE2ESuite.scala]: Just did :)

File: connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/SparkSessionE2ESuite.scala
[the remark this answers is not in the document]
Comment (cdkrot) [reply connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/SparkSessionE2ESuite.scala:1]: < 5 seconds

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/connect/client/Heartbeat.scala [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+    extends Heartbeat
+    with Logging {
+  val beatLevel: AtomicInteger = new AtomicInteger(0)
+  var count: AtomicLong = new AtomicLong(0)
+
+  override def heartbeatLevel: Int = beatLevel.get()
+
+  override def pingCount: Long = count.get()
+
+  private val pingThread = new Thread() {
```
Comment (hvanhovell) [inline connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/connect/client/Heartbeat.scala:59]: Why not use an ExecutorService and schedule a runnable?

Comment (cdkrot) [reply connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/connect/client/Heartbeat.scala:59]: Sure, I will look into it

**Pull request closed without merging**